src/tls.jl

```diff
@@ -0,0 +1,172 @@
+struct TaskLocalState
+    device::ROCDevice
+    context::HIPContext
+    queues::Vector{Union{ROCQueue,Nothing}}
+    streams::Vector{Union{HIPStream,Nothing}}
+    priority::Symbol
+end
+function TaskLocalState(device::Union{ROCDevice,Nothing},
+                        context::Union{HIPContext,Nothing},
+                        queue::Union{ROCQueue,Nothing},
+                        stream::Union{HIPStream,Nothing},
+                        priority::Symbol)
+    if device === nothing
+        if queue === nothing
+            device = Runtime.get_default_device()
+            queue = ROCQueue(device; priority, pooled=true)
+        else
+            device = AMDGPU.device(queue)
+        end
+    else
+        if queue === nothing
+            queue = ROCQueue(device; priority, pooled=true)
+        else
+            @assert queue.priority == priority
+        end
+    end
+    if context === nothing
+        context = HIPContext(device_id(device))
+    end
+    HIP.context!(context)
+    if stream === nothing
+        stream = HIPStream()
```

Comment:
Otherwise, querying for default device in `HIPStream()` causes stackoverflow.

```suggestion
        stream = HIPStream(device)
```